internal/envoy/v3/route.go

```diff
@@ -402,3 +411,123 @@ func containsMatch(s string) *envoy_route_v3.HeaderMatcher_SafeRegexMatch {
 		SafeRegexMatch: SafeRegexMatch(regex),
 	}
 }
+
+func CookieRewriteConfig(policies []dag.CookieRewritePolicy) *any.Any {
+	codeTemplate := `
+function envoy_on_response(response_handle)
+	rewrite_table = {}
+
+	{{range $i, $p := .}}
+	function cookie_{{$i}}_attribute_rewrite(attributes)
+		response_handle:logDebug("rewriting cookie \"{{$p.Name}}\"")
+
+		{{if $p.Path}}attributes["Path"] = "Path={{$p.Path}}"{{end}}
+		{{if $p.Domain}}attributes["Domain"] = "Domain={{$p.Domain}}"{{end}}
+		{{if $p.SameSite}}attributes["SameSite"] = "SameSite={{$p.SameSite}}"{{end}}
+		{{if eq $p.Secure 1}}attributes["Secure"] = nil{{end}}
+		{{if eq $p.Secure 2}}attributes["Secure"] = "Secure"{{end}}
+	end
+	rewrite_table["{{$p.Name}}"] = cookie_{{$i}}_attribute_rewrite
+	{{end}}
+
+	function rewrite_cookie(original)
+		local original_len = string.len(original)
+		local name_end = string.find(original, "=")
+		if name_end == nil then
+			return original
+		end
+		local name = string.sub(original, 1, name_end - 1)
+
+		local rewrite_func = rewrite_table[name]
+		-- We don't have a rewrite rule for this cookie.
+		if rewrite_func == nil then
+			return original
+		end
+
+		-- Find cookie value via ; or end of string
+		local value_end = string.find(original, ";", name_end)
+		-- Save position to use as iterator
+		local iter = value_end
+		if value_end == nil then
+			-- Set to 0 since we have to subtract below if we did find a ;
+			value_end = 0
+			iter = original_len
+		end
+		iter = iter + 1
+		local value = string.sub(original, name_end + 1, value_end - 1)
+
+		-- Parse original attributes into table
+		-- Keyed by attribute name, values are <name>=<value> or just <name>
+		-- so we can easily rebuild, esp for attributes like 'Secure' that
+		-- do not have a value
+		local attributes = {}
+		while iter < original_len do
+			local attr_end = string.find(original, ";", iter)
+			local new_iter = attr_end
+			if attr_end == nil then
+				-- Set to 0 since we have to subtract below if we did find a ;
+				attr_end = 0
+				new_iter = original_len
+			end
+			local attr_value = string.sub(original, iter + 1, attr_end - 1)
+			-- Strip whitespace from front
+			attr_value = string.gsub(attr_value, "^%s*(.-)$", "%1")
```

Comment:
Looking at [RFC6265](https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1), it seems to me that attribute values must be always separated exactly by single space

```
set-cookie-string = cookie-pair *( ";" SP cookie-av )
```

but maybe there are misbehaving servers that add multiple spaces or none and stripping whitespace fixes that?

Reply:
yeah, might make sense to be stricter and reject rewriting illegally structured set-cookie headers or assume valid content and act accordingly (though the latter could be odd/cause issues), this was an attempt to be a little lenient which might be wrong haha, might need more tests to see if we can break what I've written in practice as well